### ISSUE TITLE
Glide 4.8.0-SNAPSHOT doesn't work with AndroidX

### ISSUE THREAD

**Issue by Sander0542** [bumptech/glide#3184]
I have downloaded Glide 4.8.0-SNAPSHOT, but when I try to compile my app I get this error. This was after I upgraded to AndroidX

```
    [javac] /home/travis/build/Sander0542/Makeroid/appinventor/components/src/com/google/appinventor/components/runtime/util/MakeroidUtil.java:73: error: cannot access FragmentActivity
    [javac]             Glide.with(context).load(path).apply(new RequestOptions().override(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL)).into(view);
    [javac]                  ^
    [javac]   class file for android.support.v4.app.FragmentActivity not found
```

**Comment by gabrielfeo** [comment]
I'm having the same issue even when using 4.8.0-SNAPSHOT. I thought #3139 had fixed this already.

If I declare a `androidx.annotation:annotation:1.0.0-alpha3` dependency, or nothing at all, Glide gives errors because it can't find the annotation classes on the old namespace:

```
error: compiler message file broken: key=compiler.err.Processor: org.jetbrains.kotlin.kapt3.ProcessorWrapper@1f3f57ab arguments={0}, {1}, {2}, {3}, {4}, {5}, {6}, {7}
error: cannot access NonNull   class file for android.support.annotation.NonNull not found
error: compiler message file broken: key=compiler.err.Processor: org.jetbrains.kotlin.kapt3.ProcessorWrapper@1f3f57ab arguments={0}, {1}, {2}, {3}, {4}, {5}, {6}, {7}
error: cannot access Nullable   class file for android.support.annotation.Nullable not found
```


If I declare, however, a `com.android.support:support-annotations:28.0.0-alpha3` dependency, Glide gives errors for not finding the `FragmentActivity` class on the old namespace, as well as some `GlideOptions` issue:

```
error: compiler message file broken: key=compiler.err.Processor: org.jetbrains.kotlin.kapt3.ProcessorWrapper@4820a609 arguments={0}, {1}, {2}, {3}, {4}, {5}, {6}, {7}
error: cannot access FragmentActivity   class file for android.support.v4.app.FragmentActivity not found
[kapt] An exception occurred: java.lang.RuntimeException: javax.annotation.processing.FilerException: Attempt to recreate a file for type com.gabrielfeo.test.glide4.GlideOptions
```

**Comment by dougnazar** [comment]
Just spent the morning trying to figure this out. 

The Glide aar/jar files don't natively support androidx. If you're not using Android Studio or not enabling Jetifier, you'll have to use pre-converted aar/jar files. Easiest is a quick Android Studio project with Jetifier, then look for jetified-{filename}.

If you're using Android Studio w/Jetifier, #3139 just adds support for the annotation processor only. However, in order for it to detect you're using androidx, you need to import the androidx annotations into the correct namespace. The following is working for me.

```
    implementation 'com.github.bumptech.glide:glide:4.8.0-SNAPSHOT'
    annotationProcessor 'androidx.annotation:annotation:1.0.0-beta01'
    annotationProcessor 'com.github.bumptech.glide:compiler:4.8.0-SNAPSHOT'

```

**Comment by colinrtwhite** [comment]
Do you have the Jetifier enabled (add `android.enableJetifier=true` to your `gradle.properties`)? That should convert Glide's support lib dependencies to be Android X compatible.

**Comment by Sander0542** [comment]
@dougnazar so you mean that I need to compile the code of Glide with Jetifier enabled?

**Comment by dougnazar** [comment]
Not quite. Jetifier is a utility in Android Studio 3.2+ which will take existing libraries using the old support libraries and translate them to using androidx libraries. This happens transparently if you're compiling with Android Studio (I think anything using the Android Gradle plugin) with the following settings in your gradle.properties file:

    android.enableJetifier=true
    android.useAndroidX=true

 If you're not using Android Studio to build your project, you'll need to use an already converted version of the libraries. You can do this by:

1. Create an empty androidx project in Android Studio. 
2. Add imports for Glide to your project. 
3. Compile. 

You should now have a file '**jetified-glide-4.8.0-SNAPSHOT.aar**' (and the others) in your gradle cache directory. Replace the original Glide libraries with the jetified versions in your project.

What Jetifier won't do is translate your own code and since an annotation processor basically adds code for you behind the scenes, #3139 enabled the Glide annotation processor (and only the processor) to detect if you are using androidx and emit the correct helper code.

**Comment by gabrielfeo** [comment]
I have enabled the AndroidX and Jetifier flags in `gradle.properties` but I'm still getting the same errors I was getting before. I am using `kapt` configuration though. I noticed you're using Java with `annotationProcessor`.

**Comment by dougnazar** [comment]
I haven't played with Kotlin much, but a quick test seems to work. Just replacing `annotationProcessor` with `kapt`, I see my `applyOptions` being called.

```
implementation 'com.github.bumptech.glide:glide:4.8.0-SNAPSHOT'
kapt 'androidx.annotation:annotation:1.0.0-beta01'
kapt 'com.github.bumptech.glide:compiler:4.8.0-SNAPSHOT'
```
```
@GlideModule
class MyAppGlideModule : AppGlideModule() {

    override fun applyOptions(context: Context, builder: GlideBuilder) {
        Log.d("Test", "applyOptions");
    }
}
```

**Comment by dougnazar** [comment]
@Sander0542 , I'm not sure what more I can add. If you're using Android Studio, the steps are basically the glide install docs.

1. Create an empty activity application.
2. Add Glide snapshot repository.
```
    maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
```
3. Add Glide imports.
```    
    implementation 'com.github.bumptech.glide:glide:4.8.0-SNAPSHOT'
    annotationProcessor 'androidx.annotation:annotation:1.0.0-beta01'
    annotationProcessor 'com.github.bumptech.glide:compiler:4.8.0-SNAPSHOT'
```
4. Optionally add an AppGlideModule.
5. Add an ImageView.
6. Use Glide to load an image.
7. Compile & Run.

**Comment by Sander0542** [comment]
@dougnazar I created a empty Android Studio project with Glide, but I cant find the file (jetified-glide-4.8.0-SNAPSHOT.aar) in my gradle cache.

**Comment by sjudd** [comment]
To make sure I'm clear, the remaining task for Glide is to convert references in the library itself to the androidx versions. That can be done automatically via jetifier for now, though it's not totally trivial.

The problem with migrating to androidx is that we would break anyone still using the old support libraries in their applications. If we don't migrate, users who use androidx have a workaround. 

It probably doesn't make sense to consider migrating until androidx comes out of alpha and we think that more people are using androidx than android.support.

**Comment by mufumbo** [comment]
@sjudd for most Android developers (give me a thumbs up if you think like this) waiting it to come out of alpha doesn't make any sense, because: 
- if I want to stop  progress and keep old support libs, it would be trivial to Target an old version of glide on my Gradle conf (it would be already done and I wouldn't have to do anything). This will happen to all the other libs. Doing this is easier than all the BS people need to do for using glide with androidx.
- new glide users will be puzzled and will just pick another library, since they won't be able to compile glide with androidx. Most of the new documentation already points to androidx. They will think glide is outdated and won't work with the new stuff.
- support lib is Google backed, so probably will never leave alpha. Also, it's pretty stable already.
- hell, I use glide since the first version and  now am thinking its outdated and considering switching.

Better approach:
- update the major of glide to 5.0.0 and make it only androidx compatible. I don't see Google rolling back anytime soon.
- for hell sake, call it 5.0.0-alpha if you want.
- for major bugs you can cherry pick back to version 4, but new developments should happen on androidx 

But please allow people to include glide with androidx directly from Glade. It makes no sense that a great lib like this will die out because of such a simple update.

**Comment by mufumbo** [comment]
@TWiStErRob I still don't get that philosophy. 

Suppose I'm a developer who won't upgrade to androix (willing to stay with old libraries for the sake of stability). Why would I want to upgrade glide to the latest major version? Wouldn't I want to keep also glide with the old version that works well with my old libraries?

IMO this use case is very covered by just keeping the old glide major version down. While new developers (most people following google jetpack tutorials on the internet) won't suffer by copying and pasting old glide examples.

Anyway, just saw 4.8.0! Will try right now.. thanks!

**Comment by kagile** [comment]
Please Add support for Android X now

**Comment by svenjacobs** [comment]
Since AndroidX is stable as of 21st September and Android Studio 3.2 has also been released, please consider migrating the next version of Glide to AndroidX. Maybe you can consider making it a new major release? Then you can still backport fixes to the 4.x branch if necessary. Software development is about cutting ties to the old from time to time. Users of Glide who can't or don't want to migrate to AndroidX have to stay on 4.x then. Life is tough 😉 But not natively supporting the latest kind of Android development is a questionable decision _imho_.

**Comment by sjudd** [comment]
There's some progress on this internally. It's complex there as well. I'll update this if I ever get a better time frame.

**Comment by gcrescent** [comment]
I have a problem with Glide.with(androidx.fragment.app.Fragment) too.
After hours of trying, this solution works for me https://stackoverflow.com/a/52921362/4790007

It seems like Jetifier haven't convert to AndroidX, just remove your .idea/libraries folder, then File > Sync Project with Gradle Files. Hope it helps

**Comment by sjudd** [comment]
Working on this in a branch, still need to test it internally: https://github.com/sjudd/glide/tree/android_x

**Comment by TWiStErRob** [comment]
@sjudd have you considered publishing an artifact that is squeezed through jetifier-standalone or the AAR built with jetifier before publishing? This way the normal artifact stays as is and AndroidX-dependents can pull in the new dependency. This is of course as a temporary workaround until the AndroidX branch can be merged.

**Comment by sjudd** [comment]
@TWiStErRob good idea, but I was close enough that I just kept going. 

I've migrated to a new way of migrating code out to GitHub which will be easier for me to maintain, allow external developers to get faster updates, and made it possible to migrate to androidx on GitHub will still using some of the older support libraries internally.

This should be fixed by #3678. It would help if people could try the [latest snapshot][1] and make sure that jetifier is no longer necessary. 

[1]: http://bumptech.github.io/glide/dev/snapshots.html
